draco/synthesis/mockcatalog.py

```diff
@@ -0,0 +1,1045 @@
+            selfunc_local = selfunc.map[:, 0, :]
+
+            # Multiply selection function into density
+            pdf = mpiarray.MPIArray.wrap(rho * selfunc_local, axis=0)
+
+        else:
+            pdf = mpiarray.MPIArray.wrap(rho, axis=0)
+
+        # Normalize by redshift weights
+        pdf = mpiarray.MPIArray.wrap(
+            pdf / np.sum(pdf, axis=1)[:, np.newaxis] * z_weights[:, np.newaxis], axis=0
+        )
+
+        # Make container for PDF
+        pdf_map = containers.Map(
+            nside=source_map.nside,
+            polarisation=False,
+            freq=source_map.index_map["freq"],
+        )
+
+        # Put computed PDF into local section of container
+        pdf_map_local = pdf_map.map[:]
+        pdf_map_local[:, 0, :] = pdf
+
+        return pdf_map
+
+    def process(self):
+        raise NotImplementedError(f"{self.__class__} must define a process method.")
+
+
+class PdfGeneratorUncorrelated(PdfGeneratorBase):
+    """Generate uniform PDF for making uncorrelated mocks."""
+
+    def process(self, source_map):
+        """Make PDF map with uniform z weights and delta_g=0.
+
+        Parameters
+        ----------
+        source_map : :class:`containers.Map`
+            Overdensity map that determines z and angular resolution
+            of output PDF map.
+
+        Returns
+        -------
+        pdf_map : :class:`containers.Map`
+            Output PDF map.
+        """
+
+        # Get local section of source map, and set to zero
+        source_map_local = source_map.map[:, 0, :]
+        source_map_local[:] = 0
+
+        # Get local and global shape of frequency axis
+        ls = source_map.map.local_shape[0]
+        gs = source_map.map.global_shape[0]
+
+        # Set each frequency channel to have equal total probability
+        z_weights = mpiarray.MPIArray.wrap(1 / gs * np.ones(ls), axis=0)
+
+        # Create PDF map
+        pdf_map = self.make_pdf_map(source_map, z_weights)
+
+        return pdf_map
+
+
+class PdfGeneratorWithSelfunc(PdfGeneratorBase):
+    """Generate PDF that incorporates a selection function."""
+
+    def process(self, source_map, selfunc):
+        """Make PDF map that incorporates the selection function.
+
+        Parameters
+        ----------
+        source_map : :class:`containers.Map`
+            Overdensity map that determines z and angular resolution
+            of output PDF map.
+        selfunc : :class:`containers.Map`
+            Selection function map. Must have same z and angular resolution
+            as source_map. Typically taken from `ResizeSelFuncMap`.
+
+        Returns
+        -------
+        pdf_map : :class:`containers.Map`
+            Output PDF map.
+        """
+
+        # Get MPI comm
+        comm_ = source_map.comm
+
+        # Get local section of selection function
+        selfunc_local = selfunc.map[:, 0, :]
+
+        # Generate weights for distribution of sources in redshift:
```

Comment:
Not yet. But Anja's changes will help with this. I think then you'll be able to do:
```
z_weights = selfunc_local.sum(axis=1)
z_weights_sum = self.comm.allreduce(z_weights.sum())
```

Actually, I think you can already do that. Things are easier if you use the Python object based `allreduce`. this probably relies on all ranks having at least one redshift.